`azurerm_service_fabric_managed_cluster` - Fix argument (networkSecurityRules.priority and networkSecurityRules.direction) contains duplicate items: 1000 and inbound issue

sinbai opened the pull request:
The purpose of this PR:

- Fix issue [#16206](https://github.com/hashicorp/terraform-provider-azurerm/issues/16206) - Per the [Doc](https://docs.microsoft.com/en-us/azure/virtual-network/network-security-groups-overview#security-rules), _You may not create two security rules with the same priority and direction._ So, currently when multiple lb rules are set, hardcoding the priority to 1000 is not feasible. Go through the [NSG rules guidance](https://docs.microsoft.com/en-us/azure/service-fabric/how-to-managed-cluster-networking#nsg-rules-guidance)  and [NetworkSecurityRule  ](https://docs.microsoft.com/en-us/azure/templates/microsoft.servicefabric/managedclusters?tabs=bicep#networksecurityrule),  expose the Network Security Rules to support user to set value for it.
- Add multiple `lb_rule` test case.

Test Results:

> PASS: TestAccServiceFabricManagedCluster_basic (2869.35s)
> PASS: TestAccServiceFabricManagedCluster_nsRules (2152.67s)
> PASS: TestAccServiceFabricManagedCluster_multipleLBRules (2093.14s)
> PASS: TestAccServiceFabricManagedCluster_importError (2316.34s)
> PASS: TestAccServiceFabricManagedCluster_full (7106.22s)

katbyte commented on internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go:
```diff
@@ -789,7 +789,7 @@ func expandClusterProperties(model *ClusterResourceModel) *managedcluster.Manage
 				Direction:                  managedcluster.DirectionInbound,
 				Name:                       fmt.Sprintf("rule%d-allow-fe", rule.FrontendPort),
-				Priority:                   1000,
+				Priority:                   1000 + int64(idx),
```
instead of doing this should we maybe just expose the priority so the user can set it instead of making an assumption for them?

sinbai replied:
@katbyte  Sure, thanks for your time. I would like to confirm that only the `Priority` of NetworkSecurityRule needs to be exposed? Currently, all of properties of NetworkSecurityRule are hardcoded.

katbyte replied:
is there a reason for that and not exposing them to a user?

sinbai replied:
Thank you for your reply. Sorry for the late response. I don't know why they are not exposed to the user. This resource was contributed by koikonom in [PR](https://github.com/hashicorp/terraform-provider-azurerm/pull/14131). Maybe he has more knowledge in this field.

katbyte replied:
well regardless we should expose this one? as well name and direction? these seem to make sense?

they will have to be computed

sinbai replied:
> well regardless we should expose this one? as well name and direction? these seem to make sense?
> 
> they will have to be computed

I don't quite understand what you mean, could you explain?

katbyte replied:
could we expose these properties in schema, ie add priority with a default of 1000m and computed.

sinbai replied:
@katbyte thank you for your reply. Go through the [NSG rules guidance](https://docs.microsoft.com/en-us/azure/service-fabric/how-to-managed-cluster-networking#nsg-rules-guidance) and [NetworkSecurityRule ](https://docs.microsoft.com/en-us/azure/templates/microsoft.servicefabric/managedclusters?tabs=bicep#networksecurityrule), I assume that it makes more sense to expose all the properties instead of some properties of NSRules to user, WDYT? I have updated the code, could you give me some suggestions?

katbyte commented on internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go:
```diff
@@ -1231,3 +1264,83 @@ func lbRulesSchema() *pluginsdk.Schema {
+					Type:     pluginsdk.TypeString,
+					Required: true,
+				},
+				"priority": {
+					Type:         pluginsdk.TypeInt,
+					Required:     true,
+					ValidateFunc: validation.IntBetween(1000, 3000),
+				},
+				"protocol": {
+					Type:     pluginsdk.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						string(managedcluster.NsgProtocolAh),
+						string(managedcluster.NsgProtocolEsp),
+						string(managedcluster.NsgProtocolHttp),
+						string(managedcluster.NsgProtocolHttps),
+						string(managedcluster.NsgProtocolIcmp),
+						string(managedcluster.NsgProtocolTcp),
+						string(managedcluster.NsgProtocolUdp),
+					}, false),
+				},
+
+				"source_port_ranges": {
+					Type:     pluginsdk.TypeSet,
+					Required: true,
+					Elem:     &pluginsdk.Schema{Type: pluginsdk.TypeString},
+					Set:      pluginsdk.HashString,
+				},
+				"description": {
+					Type:         pluginsdk.TypeString,
+					Optional:     true,
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+
+				"source_address_prefixes": {
+					Type:     pluginsdk.TypeSet,
+					Optional: true,
+					Elem:     &pluginsdk.Schema{Type: pluginsdk.TypeString},
+					Set:      pluginsdk.HashString,
+				},
```
can we set these to the defaults that we were setting in the code? otherwise this will be a breaking change

katbyte replied:
```Access:                     managedcluster.AccessAllow,
				SourceAddressPrefixes:      &[]string{"0.0.0.0/0"},
				SourcePortRanges:           &[]string{"1-65535"},
				DestinationPortRanges:      &[]string{fePortStr},
				DestinationAddressPrefixes: &[]string{"0.0.0.0/0"},
				Direction:                  managedcluster.DirectionInbound,
				Name:                       fmt.Sprintf("rule%d-allow-fe", rule.FrontendPort),
				Priority:                   1000,

sinbai replied:
@katbyte thanks for your feedback. I would like to confirm the following things:

-  Do you means that all required properties in `ns_rule` should be changed to `Optional` and set default values for them to prevent breaking change? If yes, I would like to confirm only `access`, `direction`,`priority` could be set default value in schema in my knowledge, correct? For the other properties, is there a way to set default value in schema?
- Per the [doc](https://docs.microsoft.com/en-us/azure/service-fabric/how-to-managed-cluster-networking#nsgrules), even if [networkSecurityRules](https://github.com/Azure/azure-rest-api-specs/blob/a436672b07fb1fe276c203b086b3f0e0d0c4aa24/specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/stable/2022-01-01/managedcluster.json#L1135) (a.k.a `ns_rule` in terraform) is not set in API, the priority range 3001 to 4000  network security rules are added automatically to make configurations quick and easy by defalut when creating service fabric managed clusters. However, those rules could be overridden by adding custom NSG rules ( networkSecurityRules in API a.k.a `ns_rule` in terraform) in priority range 1000 to 3000. Actually, the networkSecurityRules and [loadBalancingRules](https://github.com/Azure/azure-rest-api-specs/blob/a436672b07fb1fe276c203b086b3f0e0d0c4aa24/specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabric/stable/2022-01-01/managedcluster.json#L1124) are not in a one-to-one relationship as written in the previous code. And there is no default value for custom NSG rules in the API. It is optional. So I assume that terraform adds default custom NSG rules that don't follow API behavior. Exposing `ns_rule` as an optional allows users to set rules on demand instead of meaningless defaults, which is what terraform should do, WDYT?

katbyte replied:
@sinbai - we are now allowing the user to set 
```
SourceAddressPrefixes:      &[]string{"0.0.0.0/0"},
				SourcePortRanges:           &[]string{"1-65535"},
				DestinationPortRanges:      &[]string{fePortStr},
				DestinationAddressPrefixes: &[]string{"0.0.0.0/0"},

```
- these were implied defaults, so now that we changed them to optional, thse new properties need to have those as the defaults. 
maybe even for default given users may be depending on that behaviour?

sinbai replied:
Fixed.

katbyte reviewed (changes requested):
Thanks @sinbai - exposing them all like this is exactly what i mean, now we just have to ensure that they default to what the code was doing to prevent breaking changes

tombuildsstuff commented on internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go:
```diff
@@ -902,27 +981,53 @@ func expandClusterProperties(model *ClusterResourceModel) *managedcluster.Manage
+					DestinationPortRanges:      &[]string{fePortStr},
+					DestinationAddressPrefixes: &[]string{"0.0.0.0/0"},
+					Direction:                  managedcluster.DirectionInbound,
+					Name:                       fmt.Sprintf("rule%d-allow-fe", frontendPort),
+					Priority:                   1000 + int64(idx),
```
these need to be user configurable

sinbai replied:
@tombuildsstuff thanks for your reply. Yes, the Priority has been exposed to the user as follows. This is the default value that prevents breaking change when not set by the user.
![image](https://user-images.githubusercontent.com/39109137/196314651-c6fc73ad-282a-4188-b0fe-559343816d74.png)

katbyte replied:
those are network security rules, not load balancer rules? to fix this we need to make it uer configurable, and then default them to these only values (only for lb rules) - likely in schema. maybe i'm missing something but you've not made all these properties setable?

katbyte commented on internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go:
```diff
@@ -202,6 +216,9 @@ func (k ClusterResource) Arguments() map[string]*pluginsdk.Schema {
 		},
 		"lb_rule": lbRulesSchema(),
+
+		"ns_rule": nsRulesSchema(),
```
can we change this to network_security_rules

sinbai replied:
> can we change this to network_security_rules

Fixed:).

mbfrahry reviewed (changes requested):
Hi @sinbai. I was testing this locally and saw that the `network_security_rules` aren't being deleted once removed from the config file.

mbfrahry commented:
This might take a bit more thought but from a user and maintainer perspective, it might be better to make this a separate resource. What do you think?

sinbai commented:
> 



> This might take a bit more thought but from a user and maintainer perspective, it might be better to make this a separate resource. What do you think?

@mbfrahry thanks for your feedback. It is okey for me to create a new resource separately. However, before that, I would like to summarize the story of this PR, at the same time explain and confirm the following things:

The purpose of this PR is to fix issue #16206. The root cause of issue 16206 is that the value of `priority` is hardcoded to 1000 in the code([here ](https://github.com/hashicorp/terraform-provider-azurerm/blob/8f1fde661d2afacd59fd87b19d260bf55c9a42d7/internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go#L917)is the original code about the default value setting fo NetworkSecurityRule). I assume that this is a bug in Terraform for the following reasons:

- In fact, when creating resource `azurerm_service_fabric_managed_cluster`, the API does not set default value for NetworkSecurityRule if it is not specified.
- Per the [doc](https://learn.microsoft.com/en-us/azure/virtual-network/network-security-groups-overview#security-rules), "_You may not create two security rules with the same priority and direction_".

According to the above points, there are two issues with the original code. First, a default value is set for NetworkSecurityRule. Second, set a fixed default value for the `priority`(hardcoding the priority to 1000 is not allowed when multiple loadbalancer rules are set).

Solution: 
Initially, I fixed issue #16206 by changing the fixed value of the `priority` from 1000 to dynamically plus 1 based on the number of loadbalancer. For this fix, as suggested by Katbyte, make NetworkSecurityRule as user-configurable item, expose the NetworkSecurityRule property and set it with default value like the original code to avoid breaking changes.  It make sense for me. I updated the code (the latest commit in this PR). A few notes about my fix:

- Because "Default is not valid for lists or sets" in Teraform schema, for sub-properties of typeset in NetworkSecurityRule, default values could only be set in the code to ensure that there are no breaking changes. This also means that the property `network_security_rules` should be computed. That's why network_security_rules could not be deleted.
- Explain my fix compared to original code:

1. [Here's ](https://github.com/hashicorp/terraform-provider-azurerm/blob/8f1fde661d2afacd59fd87b19d260bf55c9a42d7/internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go#L826)the expandClusterProperties function in the original code.
2. [Here's ](https://github.com/hashicorp/terraform-provider-azurerm/blob/f2fe1d1685f0bc849a6989565cc97e9b6e1e7c50/internal/services/servicefabricmanaged/service_fabric_managed_cluster_resource.go#L902)the updated code for the expandClusterProperties function in this PR.
3. Below is a comparison screenshot.
![image](https://user-images.githubusercontent.com/39109137/208395586-f49febdd-e9ef-4077-aa29-e7435c28b47a.png)

Questions:

- Could you let me know why my fix is not good as creating a separate resource?
- If creating a new resource, how to fix issue #16206 for azurerm_service_fabric_managed_cluster without breaking changes?

mbfrahry commented:
Hey @sinbai, thank you for giving me more information on this PR. I believe this PR would work but I did do a bit more investigation and I don't think this solution or the solution I proposed above is ideal anymore. 

First, I don't believe the solution proposed here is ideal as it will cause unintended consequences with Network Security Rules. For example, if someone specifies one load balancer rule and no network security rules, then we will add a network security rule for them. If they add one later down the line, we'll overwrite that network security rule with whatever they've specified. This isn't ideal because we've changed the resource in a way that they may not have wanted and didn't even know about because they don't know we're creating network security rules for them.

That last point leads me to my main concern that we are creating network security rules without the person knowing about them. I've spun up the resource with load balancing rules and no network security rules and it works just fine. I've also looked at what the portal is returning when adding a load balancing rule and it is also not adding network security rules. Do you mind confirming that that we don't need network security rules tied to load balancing rules? If that is the case, then I believe we can actually just remove network security rules from this resource entirely. 

The portal doesn't let you manage network security rules and we don't have an ask for them in the provider so I think it's safe to say that we can remove that functionality and it will fix the linked issue as they won't be getting a duplicate network security rule error anymore.

sinbai commented:
> Hey @sinbai, thank you for giving me more information on this PR. I believe this PR would work but I did do a bit more investigation and I don't think this solution or the solution I proposed above is ideal anymore.
> 
> First, I don't believe the solution proposed here is ideal as it will cause unintended consequences with Network Security Rules. For example, if someone specifies one load balancer rule and no network security rules, then we will add a network security rule for them. If they add one later down the line, we'll overwrite that network security rule with whatever they've specified. This isn't ideal because we've changed the resource in a way that they may not have wanted and didn't even know about because they don't know we're creating network security rules for them.
> 
> That last point leads me to my main concern that we are creating network security rules without the person knowing about them. I've spun up the resource with load balancing rules and no network security rules and it works just fine. I've also looked at what the portal is returning when adding a load balancing rule and it is also not adding network security rules. Do you mind confirming that that we don't need network security rules tied to load balancing rules? If that is the case, then I believe we can actually just remove network security rules from this resource entirely.
> 
> The portal doesn't let you manage network security rules and we don't have an ask for them in the provider so I think it's safe to say that we can remove that functionality and it will fix the linked issue as they won't be getting a duplicate network security rule error anymore.

@mbfrahry thanks for your suggestion, I also think completely removing the network security rules from this resource is the most ideal solution. I would confirm with service team.

jackofallops commented:
> 
> @mbfrahry thanks for your suggestion, I also think completely removing the network security rules from this resource is the most ideal solution. I would confirm with service team.

Given the amount of time passed, I'm going to close this issue for now. If we get a response from the service team on this then we can look to re-open.

Thanks!
